examples/step-12/step-12.cc

```diff
@@ -253,247 +290,168 @@ namespace Step12
-      &AdvectionProblem<dim>::integrate_boundary_term,
-      &AdvectionProblem<dim>::integrate_face_term,
-      assembler);
-  }
+  // Here we see the major difference to assembling by hand. Instead of
+  // writing loops over cells and faces, the logic is contained in the call to
+  // MeshWorker::mesh_loop() and we only need to specify what should happen on
+  // each cell, each boundary face, and each interior face. These three tasks
+  // are handled by the lambda functions inside the function below.
 
-
-  // @sect4{The local integrators}
-
-  // These are the functions given to the MeshWorker::integration_loop() called
-  // just above. They compute the local contributions to the system matrix and
-  // right hand side on cells and faces.
   template <int dim>
-  void AdvectionProblem<dim>::integrate_cell_term(DoFInfo & dinfo,
-                                                  CellInfo &info)
-  {
-    // First, let us retrieve some of the objects used here from @p info. Note
-    // that these objects can handle much more complex structures, thus the
-    // access here looks more complicated than might seem necessary.
-    const FEValuesBase<dim> &  fe_values    = info.fe_values();
-    FullMatrix<double> &       local_matrix = dinfo.matrix(0).matrix;
-    const std::vector<double> &JxW          = fe_values.get_JxW_values();
-
-    // With these objects, we continue local integration like always. First, we
-    // loop over the quadrature points and compute the advection vector in the
-    // current point.
-    for (unsigned int point = 0; point < fe_values.n_quadrature_points; ++point)
-      {
-        const Tensor<1, dim> beta_at_q_point =
-          beta(fe_values.quadrature_point(point));
-
-        // We solve a homogeneous equation, thus no right hand side shows up in
-        // the cell term.  What's left is integrating the matrix entries.
-        for (unsigned int i = 0; i < fe_values.dofs_per_cell; ++i)
-          for (unsigned int j = 0; j < fe_values.dofs_per_cell; ++j)
-            local_matrix(i, j) += -beta_at_q_point *                //
-                                  fe_values.shape_grad(i, point) *  //
-                                  fe_values.shape_value(j, point) * //
-                                  JxW[point];
-      }
-  }
-
-  // Now the same for the boundary terms. Note that now we use FEValuesBase, the
-  // base class for both FEFaceValues and FESubfaceValues, in order to get
-  // access to normal vectors.
-  template <int dim>
-  void AdvectionProblem<dim>::integrate_boundary_term(DoFInfo & dinfo,
-                                                      CellInfo &info)
+  void AdvectionProblem<dim>::assemble_system()
   {
-    const FEValuesBase<dim> &fe_face_values = info.fe_values();
-    FullMatrix<double> &     local_matrix   = dinfo.matrix(0).matrix;
-    Vector<double> &         local_vector   = dinfo.vector(0).block(0);
+    typedef decltype(dof_handler.begin_active()) Iterator;
```

Comment:
I think the `decltype` doesn't help readability here. Let's just use the actual type, `typename DoFHandler<dim>::active_cell_iterator`, as we do everywhere else. Also, use `using`.
```suggestion
    using ActiveCellIterator = typename DoFHandler<dim>::active_cell_iyerator;
```